### PR TITLE
Add CoreGraphics.Drawing.UnitTests with 22 image-rendering tests.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
**WARNING**: Every test here will fail until #1124 is complete.

Right now, this test facade emits images to disk using CGBitmapContext. It works on the reference platform.

I decided to make it a standalone test binary for one reason: I think its ideal interface is multi-modal; it needs to be able to generate a corpus of reference images on disk, and it needs to be able to diff them. For this, it needs a custom EntryPoint (included), and the ability to parse command-like arguments.

The default mode will be one that loads reference images from disk and fails tests if a number of pixels differ. *That work is not yet complete.*

Ideally, the draw tests here will be able to be plugged into another module that can perhaps render them to screen, or display them as part of a UI-driven flow for comparison and demoing purposes.

Refs #1271.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1348)

### CODE REVIEW

**Comment by DHowett** [comment]
This change contains a Makefile that was copied from Foundation.UnitTests.

**PROPOSAL**
* Create UnitTests/UT.common.mk, with the following interface:

```make
# Required
UT_NAME = Foundation # Can be automatically derived.
UT_FILES = $(wildcard *.cpp) $(wildcard *.mm)

# Optional, if we want pluggable entry points.
UT_FILES += $(UT_DEFAULT_FILES)

# Optional
UT_CFLAGS = -Wall -Werror

# A space-separated list of all linked frameworks.
# Foundation will be required by default as Logging uses it.
# Therefore, it might make sense to put it in UT.common.mk.
UT_FRAMEWORKS = CoreFoundation Foundation
UT_RESOURCE_DIRS = data/ otherdata/

include ../UT.common.mk
```

* Port Foundation.UnitTests over to it.
* Do the above work on *develop*.
* Consume this after a merge from develop.

@ms-jihua This is pretty quick work; what do you think?

File: tests/UnitTests/CoreGraphics.drawing/data/.keep [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (aballway) [inline tests/UnitTests/CoreGraphics.drawing/data/.keep]: what is this for?

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+#include <TestFramework.h>
+#include <CoreGraphics/CoreGraphics.h>
+
+#include <Starboard/SmartTypes.h>
+
+namespace testing {
+class DrawTest : public ::testing::Test {
+protected:
+    static woc::unique_cf<CGColorSpaceRef> s_deviceColorSpace;
```
Comment (aballway) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.h]: nit: would prefer these be accessed similar to parameterized test variables than implicitly available (say getContext

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    CGContextRef context;
```
Comment (aballway) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.h]: should this be in a unique_cf?

**Comment by DHowett-MSFT** [comment]
I forgot to add a few test files. That's been fixed.

File: tests/UnitTests/CoreGraphics.drawing/data/.keep (hunk 1 above)
Comment (DHowett-MSFT) [inline tests/UnitTests/CoreGraphics.drawing/data/.keep]: Git does not track directories, it tracks files. This is an empty file intended to keep the data directory around.

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    SetUpContext();
+}
+
+void testing::DrawTest::TearDown() {
+    CGContextRef context = GetDrawingContext();
+
+    woc::unique_cf<CGImageRef> image{ CGBitmapContextCreateImage(context) };
+    ASSERT_NE(nullptr, image);
+
+    woc::unique_cf<CFMutableDataRef> imageData{ CFDataCreateMutable(nullptr, 1048576) };
```
Comment (ms-jihua) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]: >1048576 [](start = 77, length = 7)

would be good to comment how this number was derived

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+void testing::DrawTest::SetUpTestCase() {
+    s_deviceColorSpace.reset(CGColorSpaceCreateDeviceRGB());
+}
+
+void testing::DrawTest::TearDownTestCase() {
+    s_deviceColorSpace.release();
+}
+
+CGSize testing::DrawTest::CanvasSize() {
```
Comment (ms-jihua) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]: >CanvasSize() [](start = 26, length = 12)

would we ever care about being able to change this?

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    return g_defaultCanvasSize;
+}
+
+void testing::DrawTest::SetUp() {
+    CGSize size = CanvasSize();
+
+    _context.reset(CGBitmapContextCreate(
+        nullptr, size.width, size.height, 8, size.width * 4, s_deviceColorSpace.get(), kCGImageAlphaPremultipliedFirst));
```
Comment (ms-jihua) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]: >size.width * 4 [](start = 45, length = 14)

1) are we ever going to have tests that render to colorspaces with not 4 components?
2) if not, call we call that out in a comment?

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp (hunk 5 above)
Comment (DHowett-MSFT) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]: Actually, the UIKitMimicTest fixture does exactly that; It presents a 2x-scaled context.

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp (hunk 4 above)
Comment (DHowett-MSFT) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]:
```c
# Number derived by absolutely guessing at the max size of an image; 1MB
1048576
```

File: tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp (hunk 6 above)
Comment (DHowett-MSFT) [inline tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp]: That is worth calling out, absolutely.

**Comment by rajsesh-msft** [comment]
:shipit:

**Pull request merged**